---
title: "Retrieve RNA-seq reads"
output: html_document
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(stringsAsFactors = FALSE)
source("../util/encode_functions.R")
```

```{bash}
wget -O samples.txt "https://www.encodeproject.org/report.tsv?type=Experiment&status=released&assay_slims=Transcription&assay_slims=Transcription&replicates.library.biosample.donor.organism.scientific_name=Homo+sapiens&biosample_ontology.term_name=HepG2&assay_title=polyA+plus+RNA-seq&biosample_ontology.classification=cell+line&assay_title=total+RNA-seq&files.read_length=50&limit=all"
```


```{r}
samples <- read.table("samples.txt",
                      sep = "\t", skip = 1, header = T) %>%
  dplyr::rename(experiment_accession = Accession) %>%
  mutate(file_info = map(experiment_accession, ~ encode_file_info(.x))) %>%
  unnest(file_info) %>% 
  group_by(experiment_accession) %>%
  mutate(rep_number = as.numeric(factor(replicate))) %>%
  unite(sample_id, experiment_accession, rep_number, sep = "-")


colnames(samples)

hmm <- samples %>% dplyr::select(sample_id, accession, Assay.title,
                                Biosample.summary, md5sum, paired_end_identifier)
```


```{r}
# NExt we'll grab the fastq file info using the API.



```

